lib/ansible/modules/cloud/smartos/vmadm.py

```diff
@@ -0,0 +1,612 @@
+        required: false
+        description:
+          - Additional qemu cmdline arguments for KVM guests.
+    quota:
+        required: false
+        description:
+          - Quota on zone filesystems (in MiBs).
+    ram:
+        required: false
+        description:
+          - Amount of virtual RAM for a KVM guest (in MiBs).
+    resolvers:
+        required: false
+        description:
+          - List of resolvers to be put into C(/etc/resolv.conf).
+    uuid:
+        required: false
+        description:
+          - UUID of the VM. Can either be a full UUID or C(*) for all VMs.
+    vcpus:
+        required: false
+        description:
+          - Number of virtual CPUs for a KVM guest.
+    vnc_port:
+        required: false
+        description:
+        - TCP port to listen of the VNC server. Or set C(0) for random,
+          or C(-1) to disable.
+    zfs_io_priority:
+        required: false
+        description:
+          - IO throttle priority value relative to other VMs.
+    zpool:
+        required: false
+        description:
+          - ZFS pool the VM's zone dataset will be created in.
+requirements:
+    - python >= 2.6
+'''
+
+EXAMPLES = '''
+- name: create SmartOS zone
+  vmadm:
+    brand: joyent
+    state: present
+    alias: fw_zone
+    image_uuid: 95f265b8-96b2-11e6-9597-972f3af4b6d5
+    firewall_enabled: yes
+    indestructible_zoneroot: yes
+    nics:
+      - nic_tag: admin
+        ip: dhcp
+        primary: true
+    internal_metadata:
+      root_pw: 'secret'
+    quota: 1
+
+- name: Delete a zone
+  vmadm:
+    alias: test_zone
+    state: deleted
+
+- name: Stop all zones
+  vmadm:
+    uuid: '*'
+    state: stopped
+'''
+
+RETURN = '''
+uuid:
+    description: UUID of the managed VM.
+    returned: always
+    type: string
+    sample: 'b217ab0b-cf57-efd8-cd85-958d0b80be33'
+alias:
+    description: Alias of the managed VM.
+    returned: When addressing a VM by alias.
+    type: string
+    sample: 'dns-zone'
+state:
+    description: State of the target, after execution.
+    returned: success
+    type: string
+    sample: 'running'
+'''
+
+from ansible.module_utils.basic import AnsibleModule
+import os
+import re
+import tempfile
+import time
+try:
+    import json
+except ImportError:
+    import simplejson as json
+
+# While vmadm(1M) supports a -E option to return any errors in JSON, the
+# generated JSON does not play well with the JSON parsers of Python.
+# The returned message contains '\n' as part of the stacktrace,
+# which breaks the parsers.
+VMADM = '/usr/sbin/vmadm'
+
+def get_vm_prop(module, uuid, prop):
+    """Lookup a property for the given VM.
+    Returns the property, or None if not found."""
+    cmd = '{0} lookup -j -o {1} uuid={2}'.format(VMADM, prop, uuid)
+
+    (rc, stdout, stderr) = module.run_command(cmd)
+
+    if rc != 0:
+        module.fail_json(msg='Could not perform lookup of {0} on {1}: {2}'.
+                         format(prop, uuid, stderr))
```

Comment:
The target platform (SmartOS) does (only?) runs Python 2.7. As I mentioned here https://github.com/ansible/ansible/pull/19696#pullrequestreview-15343573 and was recommended here: https://github.com/ansible/ansible/pull/19696#pullrequestreview-15343221 I chose this particular syntax.